service/history/shard/controller_impl.go

```diff
@@ -291,6 +292,52 @@ func (c *ControllerImpl) removeShardLocked(shardID int32, expected ControllableC
 	return current
 }
 
+// shardLingerThenClose delays closing the shard for a small amount of time,
+// while watching for the shard to become invalid due to receiving a shard
+// ownership lost error. It calls AssertOwnership to probe for lost ownership,
+// but any concurrent request may also see the error, which will mark the shard
+// as invalid.
+func (c *ControllerImpl) shardLingerThenClose(ctx context.Context, shardID int32) {
+	c.RLock()
+	shard, ok := c.historyShards[shardID]
+	c.RUnlock()
+	if !ok {
+		return
+	}
+
+	startTime := time.Now()
+	timeout := util.Min(c.config.ShardLingerTimeLimit(), shardIOTimeout)
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	qps := c.config.ShardLingerOwnershipCheckQPS()
+	// The limiter must be configured with burst>=1. With burst=1,
+	// the first call to Wait() won't be delayed.
+	limiter := rate.NewLimiter(rate.Limit(qps), 1)
```

Comment:
I wonder if this loop would be simpler with a time.NewTicker instead of a rate limiter?